File: 15. 3Sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        vector<vector<int>> res;
        int i=0;
        
        while (i < nums.size())
        {
            int j = i + 1;
            int k = nums.size() - 1;
            while(j < k)
            {
                if (nums[i] + nums[j] + nums[k] == 0)
                {
                    vector <int> temp {nums[i], nums[j], nums[k]};
                    res.push_back(temp);
                    j++;
                    k--;
                    while (j > 0 && j < nums.size() && nums[j] == nums[j-1])
                        j++;
                    while (k > 0 && k < nums.size() && nums[k] == nums[k+1])
                        k--;
                }
                else if (nums[i] + nums[j] + nums[k] > 0)
                {
                    k--;
                    while (k > 0 && k < nums.size() && nums[k] == nums[k+1])
                        k--;
                }
                else
                {
                    j++;
                    while (j > 0 && j < nums.size() && nums[j] == nums[j-1])
                        j++;
                }
            }
            i++;
            while (i > 0 && i < nums.size() && nums[i] == nums[i-1])
                i++;
        }
        return res;
    }
};

File: 15. 3Sum.py
class Solution(object):
    def threeSum(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        n = len(nums)
        nums = sorted(nums)
        res = []
        
        i=0
        while i < n:
            j = i+1
            k = n-1
            while j < k:
                if nums[i] + nums[j] + nums[k] == 0:
                    res += [[nums[i], nums[j], nums[k]]]
                    j += 1
                    k -= 1
                    while j > 0 and j < n and nums[j] == nums[j-1]:
                        j += 1
                    while k > 0 and k < n-1 and nums[k] == nums[k+1]:
                        k -= 1
                elif nums[i] + nums[j] + nums[k] > 0:
                    k -= 1
                    while k > 0 and k < n-1 and nums[k] == nums[k+1]:
                        k -= 1
                elif nums[i] + nums[j] + nums[k] < 0:
                    j += 1
                    while j > 0 and j < n and nums[j] == nums[j-1]:
                        j += 1
            i += 1
            while i < n and nums[i] == nums[i-1]:
                i += 1
        return res
